src/file.rs

```diff
@@ -59,87 +74,67 @@ impl FileEntry {
     /// Note that this will return [`None`] if the datetime in the cabinet file
     /// was not a valid date/time.
     pub fn datetime(&self) -> Option<PrimitiveDateTime> {
-        self.datetime
+        self.entry.datetime
     }
 
     /// Returns the total size of the file when decompressed, in bytes.
     pub fn uncompressed_size(&self) -> u32 {
-        self.uncompressed_size
+        self.entry.uncompressed_size
     }
 
     /// Returns true if this file has the "read-only" attribute set.
     pub fn is_read_only(&self) -> bool {
-        (self.attributes & consts::ATTR_READ_ONLY) != 0
+        (self.entry.attributes & consts::ATTR_READ_ONLY) != 0
     }
 
     /// Returns true if this file has the "hidden" attribute set.
     pub fn is_hidden(&self) -> bool {
-        (self.attributes & consts::ATTR_HIDDEN) != 0
+        (self.entry.attributes & consts::ATTR_HIDDEN) != 0
     }
 
     /// Returns true if this file has the "system file" attribute set.
     pub fn is_system(&self) -> bool {
-        (self.attributes & consts::ATTR_SYSTEM) != 0
+        (self.entry.attributes & consts::ATTR_SYSTEM) != 0
     }
 
     /// Returns true if this file has the "archive" (modified since last
     /// backup) attribute set.
     pub fn is_archive(&self) -> bool {
-        (self.attributes & consts::ATTR_ARCH) != 0
+        (self.entry.attributes & consts::ATTR_ARCH) != 0
     }
 
     /// Returns true if this file has the "execute after extraction" attribute
     /// set.
     pub fn is_exec(&self) -> bool {
-        (self.attributes & consts::ATTR_EXEC) != 0
+        (self.entry.attributes & consts::ATTR_EXEC) != 0
     }
 
     /// Returns true if this file has the "name is UTF" attribute set.
     pub fn is_name_utf(&self) -> bool {
-        (self.attributes & consts::ATTR_NAME_IS_UTF) != 0
+        (self.entry.attributes & consts::ATTR_NAME_IS_UTF) != 0
     }
 }
 
-impl<'a, R: Read + Seek> Read for FileReader<'a, R> {
+impl<'a> Read for FileReader<'a> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        debug_assert!(self.offset <= self.size);
         let bytes_remaining = self.size - self.offset;
         let max_bytes = bytes_remaining.min(buf.len() as u64) as usize;
         if max_bytes == 0 {
             return Ok(0);
         }
-        let bytes_read = self.reader.read(&mut buf[0..max_bytes])?;
+
+        let reader = &mut &self.reader.inner;
+        let bytes_read = reader.read(&mut buf[0..max_bytes])?;
         self.offset += bytes_read as u64;
         Ok(bytes_read)
     }
 }
 
-impl<'a, R: Read + Seek> Seek for FileReader<'a, R> {
```

Comment:
Is it still possible to support `Seek` for `FileReader`?

Reply:
@mdsteele, yes, it's possible. But it's bad for performance... PR doesn't finish, I want to add support `Cabinet::read_file`, I think I can do it.